Build popover subcomponents

statsumi97 opened the pull request:
Worked on this on new branch:

https://trello.com/c/MffIDTs5/32-popover-build-subcomponents

MenuItem:

- Restored code from old branch
- Fixed styling to better match what's on Figma 

Divider:

- Restored code from old branch

Things to note:

- I know the icon should be different instead of copy-pasting an emoji. I couldn't remember how to use what's in the Icons component but will continue to work on this in the meantime. 

To QA: http://localhost:6006/?path=/story/menuitem--default

Commit:
Create scaffold for SectionTitle

Commit:
Add storybook scaffold for SectionTitle

Commit:
Add children logic to SectionTitle

Commit:
Styled section title and add story

statsumi97 committed:
fixed styling for MenuItem default

statsumi97 committed:
Fixed styling for Hover State

statsumi97 committed:
Fixed styling for Focus State

statsumi97 committed:
Fixed styling for disabled state

statsumi97 committed:
Restoried divider

statsumi97 committed:
Merge branch 'main' into build-popover-subcomponents

statsumi97 committed:
CI: Bump version to 1.15.0

statsumi97 committed:
Built Popover

statsumi97 committed:
Build popover stories

statsumi97 committed:
Fixed toggle issue by removing portal

statsumi97 committed:
Changed popover title

statsumi97 committed:
Adjusted styling in Popover.tsx

statsumi97 committed:
Styled arrow

statsumi97 committed:
Added SectionTitle and Divider props to Popover Stories

vuonga1103 commented on src/components/Divider/Divider.tsx:
```diff
@@ -0,0 +1,19 @@
+  variant?: 'primary' | 'secondary';
+}
+
+const Divider: React.FC<DividerProps> = () => {
```
i don't think the divider needs a variant prop bc it will always be one variant. 

we can just add type to props and prop-spread so that whatever HTML attribute we pass as props to `Divider` component gets also passed to `StyledDivider`

```javascript
const Divider = (props: React.HTMLAttributes<HTMLElement>) => {
  return <StyledDivider {...props} />;
};
```

& remove unnecessary interface `DividerProps`

vuonga1103 commented on src/components/Divider/__docs__/Divider.stories.tsx:
```diff
@@ -0,0 +1,17 @@
+const meta: Meta<typeof Divider> = {
+  title: 'Divider',
+  component: Divider,
+  argTypes: {
```
no need for argTypes since we don't really need any control or options for Divider. would revise to

```javascript
import React from 'react'
import type { Meta } from '@storybook/react';
import Divider from '../Divider';

const meta: Meta<typeof Divider> = {
  title: 'Divider',
  component: Divider,
};

export default meta;

export const Default = {
  render: (args) => <Divider {...args} />,
};
```

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,102 @@
+}) => {
+  return (
+    <StyledButton
+      className={`menu-item ${disabled ? 'disabled' : ''}`}
```
instead of using className, could we pass disabled as props into `StyledButton` and use it within a callback in the styled component?

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,102 @@
+  /* Hover state */
+  &:hover {
+    background-color: transparent; /* Set background to transparent */
+    color: var(--color-content-brand, #026fe4);
```
where possible could we pull token vars from `theme`?

vuonga1103 commented on src/components/index.ts:
```diff
@@ -1,3 +1,4 @@
 export * from './Button';
 export * from './Icon';
 export * from './Avatar';
+export * from './SectionTitle';
```
need an export for `Divider`

statsumi97 committed:
Using React.HTMLAttributes<HTMLElement> to accept any valid HTML attribute for a div

statsumi97 committed:
remove the argTypes related to variant.

statsumi97 committed:
Updated Meta type usage for removal of argTypes and Simplified story definition as suggested

statsumi97 committed:
Adjust StyledButton to accept disabled as a prop and apply styles conditionally

statsumi97 committed:
adjusted the StyledButton to correctly access the base and semantic tokens for styling

statsumi97 committed:
Merge remote changes

statsumi97 commented:
I have pushed the updated code which addresses the following: 

- Removed variant prop and unnecessary interface DividerProps
- removed argTypes
- Passed disabled as props into StyledButton
- Pulled token vars from theme
- Exported Divider, MenuItem and Popover into Index.ts

I have also pushed code for the Popover. 

Trello:
https://trello.com/c/MffIDTs5/32-popover-build-subcomponents
https://trello.com/c/QlXzcs1E/36-popover-build-popover

To QA: http://localhost:6006/?path=/story/menuitem--disabled

statsumi97 committed:
fixed disabled color

statsumi97 committed:
dynamically use icons from the Storybook component

statsumi97 committed:
use the icon's name, size, and color as arguments in MenuItem.stories.tsx

statsumi97 commented:
Pushed code which adds dynamic icons. 

For QA: http://localhost:6006/?path=/story/menuitem--disabled

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,79 @@
+  cursor: pointer;
+  border: none;
+  display: flex;
+  width: 191px;
```
setting a fixed width might give rise to visual bugs, typically you want to do 100% to fit the size of whatever container the menu item is in

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,79 @@
+
+//Adjust StyledButton to accept disabled as a prop and apply styles conditionally
+const StyledButton = styled.button<{ disabled?: boolean }>`
+  ${({ theme, disabled }) => `
```
could destructure out some theme vars so we don't have to use `theme.base....` each time we use a design token. 

could you use destructuring so that we can do... `spacing.spacing4`, `border.radius.radius4`, etc.?

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,79 @@
+  iconColor?: keyof EquityTheme['semantic']['color']['content'];
+}
+
+const MenuItem: React.FC<MenuItemProps> = ({
```
if we want to use the menu item without an icon, how would we be able to do that?

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,79 @@
+      {...props}
+      // Add event handlers for hover and focus
+      onMouseEnter={() => console.log('hover')}
+      onFocus={() => console.log('focus')}
```
might be unnecessary to add event handlers. we would ideally want to leave these for the users to configure on a case to case basis. 

additionally always make sure that `{...props}` comes last. otherwise whatever the user adds might get overridden.

vuonga1103 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+import type { Meta, StoryObj } from '@storybook/react';
```
a few storybook related things:
1. icon size control is text right now so the SB-viewer would be able to type in anything. how can we constrict this control to only allow the user to select one of the available icon sizes?

https://github.com/femmecubator/equity-ui/assets/58869812/dd83fba3-1d4b-413b-a3ca-fe74b258361c

3. could we have a story for a no-icon variation of the component? 

4. when i add text that is very long - the icon progressively gets smaller and ultimately disappears. let's address that. 

https://github.com/femmecubator/equity-ui/assets/58869812/4ddeb82f-dc4e-4a0f-9e43-17b9afb4394c

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,79 @@
+}) => {
+  return (
+    <StyledButton
+      disabled={disabled}
```
for accessibility reasons we should add `role="menuitem"`

read more here https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menuitem_role

vuonga1103 commented on src/components/Popover/Popover.tsx:
```diff
@@ -0,0 +1,70 @@
+  `}
+`;
+
+const Popover: React.FC<PopoverProps> = ({ open, anchorElement, children }) => {
```
we need to figure out how to render the popover in a portal. 

> Portals render parts of a React component tree into a different part of the DOM. This is useful for UI components that need to appear over the top of other components, such as modal dialog, flag and tooltip.

vuonga1103 replied:
https://react.dev/reference/react-dom/createPortal

vuonga1103 replied:
^ this is the part that i feel like will take longer - hence wanting to focus on the build of the subcomponents. i'd say for you to focus on addressing the other subcomponents' comments before heading to this.

vuonga1103 commented on src/components/Popover/Popover.tsx:
```diff
@@ -0,0 +1,70 @@
+  display: flex;
+  flex-direction: column;
+  align-items: flex-start;
+  width: 227px;
```
would we want the popover to have this width always? or should it be more dynamic? perhaps we can set a min-width?

vuonga1103 commented on src/components/Popover/Popover.tsx:
```diff
@@ -0,0 +1,70 @@
+  border: none;
+  box-shadow: 0px 4px 12px rgba(51, 51, 51, 0.2);
+  padding: 16px;
+  border-radius: var(--border-radius-default, 16px);
```
previous comments re: use of tokens apply here.

vuonga1103 commented on src/components/Popover/Popover.tsx:
```diff
@@ -0,0 +1,70 @@
+    return null;
+  }
+
+  return <StyledPopover position={position}>{children}</StyledPopover>;
```
what role can we assign to this to make it more accessible?

vuonga1103 commented on src/components/Popover/Popover.tsx:
```diff
@@ -0,0 +1,70 @@
+  `}
+`;
+
+const Popover: React.FC<PopoverProps> = ({ open, anchorElement, children }) => {
```
would the user be able to pass other attributes to the popover? how may they be able to do that?

vuonga1103 commented on src/components/Divider/Divider.tsx:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import styled from '@emotion/styled';
+
+// Styled divider using styled-components for styling
```
let’s remove unnecessary comments (totally understand that you use these for learning purposes, but for production code, we should use comments extremely sparingly and write code in such a way that it’d be self-explanatory). 

applies to this and wherever else in the PR

vuonga1103 reviewed (changes requested):
great job on this! requesting the following changes. i would prioritize things in the following order:
1. addressing all comments except the one about rendering popover on a portal
2. start looking into how you may want to test these components
3. work on the portal problem

feel free to comment on each of the thread below if you need additional clarification.

oh also, no need to relink the storybook localhost and trello links! :D

statsumi97 committed:
updated the width property in MenuItem.tsx

statsumi97 committed:
destructured the necessary parts of the theme object within the styled component

statsumi97 committed:
adjusted the conditional rendering logic in MenuItem.tsx to not render any icon when not provided

limoncello526 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+    iconName: 'link',
+    iconSize: 'small',
+    iconColor: 'default',
+  },
```
if you find yourself repeating codes, write a function for it

limoncello526 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+    disabled: true,
+    iconName: 'link',
+    iconSize: 'small',
+    iconColor: 'default',
```
x2

limoncello526 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+    text: 'Menu Item',
+    iconName: 'link',
+    iconSize: 'small',
+    iconColor: 'default',
```
x3

limoncello526 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+    text: 'Menu Item',
+    iconName: 'link',
+    iconSize: 'small',
+    iconColor: 'default',
```
x4

limoncello526 commented on src/components/SectionTitle/SectionTitle.tsx:
```diff
@@ -0,0 +1,26 @@
+const StyledSectionTitle = styled.h2`
+  font-weight: 700;
+  font-size: 18px;
+  line-height: 22px;
```
let's move this down and apply theme

statsumi97 committed:
Fixed styling

statsumi97 committed:
added ...props last in StyledButton

statsumi97 committed:
ensured that iconSize only accepts DefaultIconSizes in MenuItem.tsx

statsumi97 committed:
specified the control type for iconSize to restrict the options available in MenuItem stories

statsumi97 committed:
added a new story configuration that doesn't specify an iconName in MenuItem stories

statsumi97 committed:
applied a CSS solution using flex-shrink, flex-grow, and min-width properties in MenuItem.tsx

statsumi97 committed:
added role=menuitem for accessibility reasons in MenuItem.tsx

statsumi97 committed:
removed unnecessary comments throughout code

statsumi97 committed:
created a function to avoid code repetition in MenuItem stories

statsumi97 committed:
adjusted the styling to use the theme in SectionTitle.tsx

statsumi97 committed:
Added tests to MenuItem.test

vuonga1103 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,64 @@
+
+type IconColor = keyof EquityTheme['semantic']['color']['content'];
+
+function createMenuItemArgs({
```
use `menuItemArgs` object in place of function

vuonga1103 commented on src/components/MenuItem/__docs__/MenuItem.stories.tsx:
```diff
@@ -0,0 +1,64 @@
+  text = 'MenuItem',
+  iconName = 'link' as IconName,
+  iconSize = 'small' as DefaultIconSizes,
+  iconColor = 'default' as IconColor,
```
add `noIcon` prop to deal with no icons

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,109 @@
+  iconSize = 'large',
+  iconColor = 'default',
+  disabled = false,
+  onMouseEnter = () => console.log('hover'),
```
can remove handler props

vuonga1103 commented on src/components/MenuItem/MenuItem.test.tsx:
```diff
@@ -0,0 +1,44 @@
+import MenuItem from '.';
+import { theme } from '../../theme';
+
+const customRender = (
```
use `contextRender`

vuonga1103 commented on src/components/MenuItem/MenuItem.test.tsx:
```diff
@@ -0,0 +1,44 @@
+};
+
+describe('MenuItem component', () => {
+  it('should render correctly', () => {
```
add render tests for other subcomponents

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,109 @@
+      align-items: flex-start;
+      border-radius: ${radius4};
+      color: ${disabled ? gray60 : gray90};
+      /* body/small */
```
unnecessary comment

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,109 @@
+      &:focus {
+        color: ${disabled ? gray30 : gray90};
+        outline: ${disabled ? 'none' : `1px solid ${gray90}`};
+        padding-right: 95px;
```
let's remove this hardcode of 95px

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,109 @@
+  size: DefaultIconSizes;
+}
+
+const MenuItem: React.FC<MenuItemProps> = ({
```
nice to have - vertical alignment of icon & text 
<img width="182" alt="image" src="https://github.com/femmecubator/equity-ui/assets/58869812/bbc7c0f9-3bdb-4fcd-b76b-3a1574e4b096">

vuonga1103 commented on src/components/MenuItem/MenuItem.tsx:
```diff
@@ -0,0 +1,109 @@
+          <Icon name={iconName} size={iconSize} color={iconColor} />
+        </IconWrapper>
+      )}
+      <span
```
nit - create styled component

statsumi97 committed:
merge

statsumi97 committed:
use menuItemArgs object in place of function

statsumi97 committed:
add noIcon prop to deal with no icons

statsumi97 committed:
add noIcon prop to deal with no icons

statsumi97 committed:
removed handler props

statsumi97 committed:
use contextRender

statsumi97 committed:
added tests to rest of the components

statsumi97 committed:
Removed unnecessary comments

statsumi97 committed:
removed hardcode of 95px

statsumi97 committed:
vertical alignment of icon & text

statsumi97 committed:
create styled component

statsumi97 committed:
CI: Bump version to 1.16.0

vuonga1103 commented on src/components/MenuItem/MenuItem.test.tsx:
```diff
@@ -0,0 +1,44 @@
+import MenuItem from '.';
+import { theme } from '../../theme';
+
+const contextRender = (
```
use our `contextRender` from src/shared/utils/contextRender.tsx

vuonga1103 commented on src/components/SectionTitle/SectionTitle.test.tsx:
```diff
@@ -0,0 +1,19 @@
+import SectionTitle from '.';
+import { theme } from '../../theme';
+
+const contextRender = (ui: React.ReactElement) => {
```
use contextRender from `src/shared/utils/contextRender.tsx`

statsumi97 committed:
use contextRender from src/shared/utils/contextRender.tsx

statsumi97 committed:
merge
